fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,573 @@
+use serde::{Deserialize, Serialize};
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::time::SystemTime;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use crate::base_storage::{BaseStorage, SyncStatus};
+use crate::monoid::Monoid;
+// use crate::utils::read_version_2_fs;
+use data_error::{ArklibError, Result};
+
+/*
+Note on `FolderStorage` Versioning:
+
+`FolderStorage` is a basic key-value storage system that persists data to disk.
+where the key is the path of the file inside the directory.
+
+
+In version 2, `FolderStorage` stored data in a plaintext format.
+Starting from version 3, data is stored in JSON format.
+
+For backward compatibility, we provide a helper function `read_version_2_fs` to read version 2 format.
+*/
+const STORAGE_VERSION: i32 = 3;
+
+/// Represents a folder storage system that persists data to disk.
+pub struct FolderStorage<K, V>
+where
+    K: Ord,
```

Comment:
Since the key in `FolderStorage` is the filename, we should add a constraint that `K` must implement `std::fmt::Display`, right?

Reply:
Good point. Technically, we could solve the task for arbitrary key type by storing the keys together with values. But this complicates the solution unnecessary for this moment. We'll be able to loose the constraint when we implement "chunked storage".

Reply:
ChunkedStorage is meant to store n key-value pairs per file. FolderStorage is a special case where n = 1. We can optimise the special case by storing the key only in the in-memory mapping. This means that the key only needs constraints to be stored in the BTreeMap which would be Ord + Eq.

Reply:
Yes, but the easiest way to persist the keys is in the filenames.

```
<storage_root>
|- qwe
|- asd
|- zxc
```

`"qwe"`, `"asd"`, `"zxc"` are the keys here, e.g. file `<storage_root>/qwe` contains value assigned to key `"qwe"`

Do we store data like this currently? We could just fix the keys type to `String` but we want to use the storage at least with numerical keys, too. So, `Display` is the most practical generalization.

If we allow keys to be just `Ord`, then we need to serialize, store, load and deserialize them. I would postpone such work till we start working on the chunked storage.